File: ForgeSwift/ForgeCore/Sources/ForgeCore/Intelligence/MetabolicWatchSignals.swift
import Foundation

/// Inputs for the Apple Watch-derived metabolic estimate.
///
/// Every value is an already-available HealthKit rollup (DailyHealthStats +
/// weekly trends) — no new HealthKit queries were added for this. A value of 0
/// (or negative) means "no data": HealthKit leaves missing fields at 0, and the
/// estimator treats them as absent rather than inventing a reading.
public struct MetabolicWatchInputs: Sendable, Equatable {
    public var restingHRBpm: Double
    public var hrvMs: Double
    public var hrvBaselineMs: Double
    public var vo2Max: Double
    public var activeCalories: Double
    public var exerciseMinutes: Double
    public var sleepHours: Double
    public var sleepBaselineHours: Double

    public init(
        restingHRBpm: Double = 0,
        hrvMs: Double = 0,
        hrvBaselineMs: Double = 0,
        vo2Max: Double = 0,
        activeCalories: Double = 0,
        exerciseMinutes: Double = 0,
        sleepHours: Double = 0,
        sleepBaselineHours: Double = 0
    ) {
        self.restingHRBpm = restingHRBpm
        self.hrvMs = hrvMs
        self.hrvBaselineMs = hrvBaselineMs
        self.vo2Max = vo2Max
        self.activeCalories = activeCalories
        self.exerciseMinutes = exerciseMinutes
        self.sleepHours = sleepHours
        self.sleepBaselineHours = sleepBaselineHours
    }
}

/// One Apple Watch-derived hint shown in the Metabolic section.
///
/// Honesty contract: `line` always names its source ("Apple Watch"), and any
/// line that interprets data (vs. merely reporting it) says "estimate". These
/// are lifestyle hints — never measurements of metabolic state, never medical
/// advice, never a diagnosis.
public struct MetabolicWatchSignal: Sendable, Equatable {
    public var id: String
    public var title: String
    public var line: String

    public init(id: String, title: String, line: String) {
        self.id = id
        self.title = title
        self.line = line
    }
}

/// The Apple Watch-derived metabolic estimate.
///
/// This exists so Metabolic Health is useful with no CGM. It is deliberately
/// NOT a score: a list of labeled, source-traceable hints plus one summary
/// line. CGM glucose remains the source of truth — `MetabolicHealthSnapshot`
/// only attaches this when there is no glucose data.
public struct MetabolicWatchEstimate: Sendable, Equatable {
    public var signals: [MetabolicWatchSignal]
    public var summaryLine: String

    public var hasSignals: Bool { !signals.isEmpty }
    public var displayLines: [String] { signals.map(\.line) }

    public static let empty = MetabolicWatchEstimate(signals: [], summaryLine: "")

    public init(signals: [MetabolicWatchSignal], summaryLine: String) {
        self.signals = signals
        self.summaryLine = summaryLine
    }

    /// Builds the estimate from watch rollups.
    ///
    /// Returns `.empty` unless at least two independent signals are present:
    /// a single data point is not a metabolic picture, and surfacing one
    /// invites over-interpretation. Zero/negative/implausible values are
    /// treated as missing, never as real readings.
    public static func evaluate(_ inputs: MetabolicWatchInputs) -> MetabolicWatchEstimate {
        var signals: [MetabolicWatchSignal] = []

        if let vo2 = plausible(inputs.vo2Max, range: 15...80) {
            signals.append(MetabolicWatchSignal(
                id: "cardio-fitness",
                title: "Cardio fitness",
                line: "≈ \(Int(vo2.rounded())) — estimated by Apple Watch"
            ))
        }

        if let rhr = plausible(inputs.restingHRBpm, range: 25...120) {
            signals.append(MetabolicWatchSignal(
                id: "resting-hr",
                title: "Resting heart rate",
                line: "\(Int(rhr.rounded())) bpm today — Apple Watch"
            ))
        }

        if let hrv = plausible(inputs.hrvMs, range: 5...300) {
            let line: String
            if let base = plausible(inputs.hrvBaselineMs, range: 5...300) {
                line = "\(Int(hrv.rounded())) ms vs your recent \(Int(base.rounded())) ms — Apple Watch estimate"
            } else {
                line = "\(Int(hrv.rounded())) ms today — Apple Watch"
            }
            signals.append(MetabolicWatchSignal(id: "hrv", title: "Heart rate variability", line: line))
        }

        let calories = inputs.activeCalories > 0 ? Int(inputs.activeCalories.rounded()) : nil
        let minutes = inputs.exerciseMinutes > 0 ? Int(inputs.exerciseMinutes.rounded()) : nil
        if calories != nil || minutes != nil {
            let bits = [
                calories.map { "\($0) active calories" },
                minutes.map { "\($0) exercise minutes" },
            ].compactMap { $0 }
            signals.append(MetabolicWatchSignal(
                id: "movement",
                title: "Movement",
                line: "\(bits.joined(separator: " · ")) today — Apple Watch"
            ))
        }

        if let sleep = plausible(inputs.sleepHours, range: 0.5...16) {
            let line: String
            if let usual = plausible(inputs.sleepBaselineHours, range: 0.5...16) {
                line = String(format: "%.1f h vs your usual %.1f h — Apple Watch estimate", sleep, usual)
            } else {
                line = String(format: "%.1f h last night — Apple Watch", sleep)
            }
            signals.append(MetabolicWatchSignal(id: "sleep", title: "Sleep", line: line))
        }

        guard signals.count >= 2 else { return .empty }

        return MetabolicWatchEstimate(
            signals: signals,
            summaryLine: "No glucose sensor connected. These are estimates from Apple Watch trends — not measurements, and not medical advice."
        )
    }
}

/// Zero/negative means "no data"; outside the range means a corrupt sample.
/// Either way the value is dropped, never presented as a reading.
private func plausible(_ value: Double, range: ClosedRange<Double>) -> Double? {
    guard value > 0, range.contains(value) else { return nil }
    return value
}

File: backend/infra/lambda/aria_core/metabolic_watch_signals.py
"""Apple Watch-derived metabolic hints when no CGM is connected.

Python port of ForgeCore's ``MetabolicWatchSignals.swift``. Honesty contract:
every line names its source, and interpreting lines say "estimate". Never a
score, never a diagnosis.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class MetabolicWatchInputs:
    resting_hr_bpm: float = 0.0
    hrv_ms: float = 0.0
    hrv_baseline_ms: float = 0.0
    vo2_max: float = 0.0
    active_calories: float = 0.0
    exercise_minutes: float = 0.0
    sleep_hours: float = 0.0
    sleep_baseline_hours: float = 0.0


@dataclass(frozen=True)
class MetabolicWatchSignal:
    id: str
    title: str
    line: str

    def to_dict(self) -> dict:
        return {"id": self.id, "title": self.title, "line": self.line}


@dataclass(frozen=True)
class MetabolicWatchEstimate:
    signals: tuple[MetabolicWatchSignal, ...]
    summary_line: str

    @property
    def has_signals(self) -> bool:
        return bool(self.signals)

    def to_dict(self) -> dict:
        return {
            "signals": [s.to_dict() for s in self.signals],
            "summaryLine": self.summary_line,
            "hasSignals": self.has_signals,
            "displayLines": [s.line for s in self.signals],
        }


EMPTY = MetabolicWatchEstimate(signals=(), summary_line="")


def _plausible(value: float, lo: float, hi: float) -> float | None:
    if value <= 0 or value < lo or value > hi:
        return None
    return value


def _swift_round(value: float) -> int:
    return int(value + 0.5) if value >= 0 else -int(-value + 0.5)


def evaluate(
    inputs: MetabolicWatchInputs,
    *,
    source: str = "Apple Watch",
) -> MetabolicWatchEstimate:
    """``source`` names the sensor. Swift callers keep ``Apple Watch``; the
    shared sidecar uses ``wearable`` so Android is not told a lie."""
    label = (source or "Apple Watch").strip() or "Apple Watch"
    signals: list[MetabolicWatchSignal] = []

    vo2 = _plausible(inputs.vo2_max, 15, 80)
    if vo2 is not None:
        signals.append(MetabolicWatchSignal(
            id="cardio-fitness",
            title="Cardio fitness",
            line=f"≈ {_swift_round(vo2)} — estimated by {label}",
        ))

    rhr = _plausible(inputs.resting_hr_bpm, 25, 120)
    if rhr is not None:
        signals.append(MetabolicWatchSignal(
            id="resting-hr",
            title="Resting heart rate",
            line=f"{_swift_round(rhr)} bpm today — {label}",
        ))

    hrv = _plausible(inputs.hrv_ms, 5, 300)
    if hrv is not None:
        base = _plausible(inputs.hrv_baseline_ms, 5, 300)
        if base is not None:
            line = f"{_swift_round(hrv)} ms vs your recent {_swift_round(base)} ms — {label} estimate"
        else:
            line = f"{_swift_round(hrv)} ms today — {label}"
        signals.append(MetabolicWatchSignal(id="hrv", title="Heart rate variability", line=line))

    calories = _swift_round(inputs.active_calories) if inputs.active_calories > 0 else None
    minutes = _swift_round(inputs.exercise_minutes) if inputs.exercise_minutes > 0 else None
    if calories is not None or minutes is not None:
        bits = []
        if calories is not None:
            bits.append(f"{calories} active calories")
        if minutes is not None:
            bits.append(f"{minutes} exercise minutes")
        signals.append(MetabolicWatchSignal(
            id="movement",
            title="Movement",
            line=f"{' · '.join(bits)} today — {label}",
        ))

    sleep = _plausible(inputs.sleep_hours, 0.5, 16)
    if sleep is not None:
        usual = _plausible(inputs.sleep_baseline_hours, 0.5, 16)
        if usual is not None:
            line = f"{sleep:.1f} h vs your usual {usual:.1f} h — {label} estimate"
        else:
            line = f"{sleep:.1f} h last night — {label}"
        signals.append(MetabolicWatchSignal(id="sleep", title="Sleep", line=line))

    if len(signals) < 2:
        return EMPTY

    return MetabolicWatchEstimate(
        signals=tuple(signals),
        summary_line=(
            f"No glucose sensor connected. These are estimates from {label} "
            "trends — not measurements, and not medical advice."
        ),
    )
